### PR TITLE
OBPIH-6103 Create endpoint to create/edit product source details

### PR DESCRIPTION
**Pull request by kchelstowski**

### CODE REVIEW

File: src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy
Comment (kchelstowski) [inline src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy:43]: Since we are about to create 3 separate endpoints for now, I've decided to go with the command approach to make sure, that when using details endpoint, a user will potentially be able only to edit details part and that the user won't be able to include any other properties, e.g. throwing a list of `productSupplierPreferences` that would be persisted if I used the `ProductSupplier` domain as a command class.

cc @awalkowiak @alannadolny @drodzewicz @jmiranda

File: grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy
Comment (kchelstowski) [inline grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:385]: as you can see, the if statement below is repeated here, and in the `saveDetails` method. I've implemented a method for that recently, while working on OBPIH-6017 (assignSourceCode): https://github.com/openboxes/openboxes/pull/4486
that we will be able to use here as soon as the feature branch is merged into develop (upgrade to grails 3.3.10) branch.

**Review by alannadolny — APPROVED** [review]

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (drodzewicz) [inline grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:37]: This looks hardcore, I think it is safe to use numbers for HTTP status codes :sweat_smile:

**Review by drodzewicz — APPROVED** [review]

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (drodzewicz) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:37]: I think I will retract my comment....

The reason I said this is that I usually dealt with HTTP codes in a number format and they are the same across all frameworks, and importing the HttpsStatus and extracting that exact number using `HttpStatus.CREATED.value()` was something that caught my attention, but thinking about this right now I think I see a point in this, this is probably easier to read and debug 

soooo... statement rollback, this looks right :+1:

File: src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy
Comment (awalkowiak) [reply src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy:43]: I wonder if we could somehow use shared constraints, like here: https://grails.github.io/legacy-grails-doc/3.3.0/ref/Constraints/Usage.html so we have a "single source of truth" regarding constraints, and won't have to change two places whenever constraints will change on the domain level.

**Review by awalkowiak — COMMENTED** [review]
I only have a question / discussion around constraints about how we want to move forward regarding sharing constraints between command and domain

File: src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy
Comment (kchelstowski) [reply src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy:43]: @awalkowiak thanks, this feels interesting. I have not known this was possible. I will check it out and get back to you.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (awalkowiak) [inline grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:34]: Is the naming `productSupplierCommand` here used on purpose (without `Details`)? If not I think it should match the command name

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (awalkowiak) [inline grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:35]: I wonder if the word `create` would not be more suitable here instead of `save`, but I'll let @jmiranda decide what he prefers here. Other than that and one minor naming thing above it LGTM.

**Review by awalkowiak — COMMENTED** [review]
Overall LGTM, just one minor thing that I'd like to hear @jmiranda's opinion.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (kchelstowski) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:34]: you are right, this is inconsistency from my side, so answering - it was not intentional.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (kchelstowski) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:35]: to include my opinion - I'd stick with `save`, since we are actually calling the db and saving the entity (and GORM's method is named `save`, not `create`, so to kinda follow the convention)
But I don't mind both, so let me know what you guys prefer.
I can also see that Grails' docs probably follow `save`/`update` pattern: https://docs.grails.org/latest/guide/REST.html

Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy]: Looking into this docs, it also suggests calling the action `save` if you do POST to the `/productSuppliers`, so then if you'd like to follow it, then you need to change the action name to `save`.

Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy]: Sounds good.

File: grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:368]:
```
ProductSupplier productSupplier = new ProductSupplier(command.properties)
if (command.hasErrors()) { 
    productSupplier.errors = command.errors
    return productSupplier
}

// do some stuff

if (!productSupplier.hasErrors()) { 
    productSupplier = productSupplierGormService.save(productSupplier)
}
return productSupplier
```

Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:368]: * should service throw exception 
* static vs compilestatic
    * @GrailsCompileStatic
    * @CompileStatic

Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:368]: @jmiranda To add something to here - I feel like we should focus on building a proper rest API, and stick to one approach or convention (in this case I think we should stick to throwing exception or error and handling it in error controller instead returning products supplier object with errors copied from command), and not mix into it different things to "satisfy" other potential use cases (calls from other services or non API controllers)

cc @kchelstowski, @drodzewicz @alannadolny

File: src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy
Comment (jmiranda) [reply src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy:38]: I found an even worse issue than what @kchelstowski was worried about. And we are definitely susceptible to this given that none of our domain or command constraints are marked as final.

https://grails.org/blog/2022-09-28-static-binding-bug.html

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: You'll also need to add one for ProductSupplierPreference.
```
        JSON.registerObjectMarshaller(ProductSupplierPreference) { ProductSupplierPreference productSupplierPreference ->
            return productSupplierPreference.toJson()
        }
```


And the supplier object in the ProductSupplier.toJson() is missing an ID, which initially prevented me from copying the JSON returned by a GET request and sending it in a PUT request.

Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: Here's my attempt at the create endpoint. The validation exception should have a better message, but I hate that I need to pass a message there.
```
    def create(ProductSupplierDetailsCommand command) {
        if (command.hasErrors()) {
            throw new ValidationException("validation errors", command.errors)
        }

        ProductSupplier productSupplier = productSupplierService.saveProductSupplier(params.id, command)

        response.status = HttpStatus.CREATED.value()

        render([data: productSupplier] as JSON)
    }
```

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:42]: And here's the update endpoint.
```
def update(ProductSupplierCommand command) {
    if (command.hasErrors()) {
        throw new ValidationException("validation errors", command.errors)
    }
    ProductSupplier productSupplier = productSupplierService.saveProductSupplier(params.id, command)

    render([data: productSupplier] as JSON)
}
```

Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:42]: I still don't like the params.id as an argument to this method, but understand why it was done that way.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:44]: remove toJson

File: grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy
Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:367]: Here's my attempt at the generic save (used for create or update). 
```
ProductSupplier saveProductSupplier(String id, ProductSupplierCommand command) {
    ProductSupplier productSupplier
    if (id == null) {
        productSupplier = new ProductSupplier(command.properties)
    }
    else {
        productSupplier = productSupplierGormService.get(id)
        if (!productSupplier) {
            throw new ObjectNotFoundException(id, ProductSupplier.class.toString())
        }
        productSupplier.properties = command.properties
    }

    if (!productSupplier.code) {
        productSupplier.code =
            identifierService.generateProductSupplierIdentifier(command?.product?.productCode, command?.supplier?.code)
    }
    return productSupplierGormService.save(productSupplier)
}
```

Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:367]: Note: I am not handling the validation on the product supplier domain object. If there's a desire to throw a validationexception from the service, you'll need to add some code here to handle it.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: The reason I would prefer to use saveProductSupplier over saveDetails or save is because the ProductSupplierService will probably also house the code to save other objects like preferences and packages that are associated with a product supplier.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:42]: Regarding the response for create and update, it is possible that we return the object as-is, but I am still working out the correct "response" for these endpoints based on best practices. There are a lot of good arguments for both sides.

**Review by jmiranda — CHANGES_REQUESTED** [review]

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (kchelstowski) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: Is there a deeper reason why you prefer Bootstrap over `toJson()` here? I think recently we've discussed that we wanted to avoid using Bootstrap and stick to `toJson()` for now even if it's not the best approach possible.

Comment (kchelstowski) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: The reason I went with `saveDetails` is for this method not to be confusing - doing it as `saveProductSupplier` in my opinion suggests, that we could update all product supplier's properties there - this is why I haven't named it like this.
But if you guys are ok with that, then I can change it.

**Comment by kchelstowski** [comment]
@jmiranda 

> And fwiw, the danger seems to exist whether I use a command or domain class since the data binding eventually occurs on the domain object. We can certainly limit the reach with a command class, but we can still do things we might not want to allow.


Yeah, this is why I've mentioned many times that I am not fan of relying on domain classes for binding or inside a command class, because such "weirds" dangerous things could happen if we rely on domain classes, not on POJOs which we have control of most often with third party libraries like MapStruct which are very flexible and configurable for any use case.

So I am not surprised that we can edit the Product and Organization, because those are existing domain object that are bound to the transaction/session and any changes on them would be tracked.
An easy way to prevent from updating them would be to either create some generic method, or just write those two lines manually after binding to the product suppleier domain:
```groovy
productSupplier.properties = command.properties
productSupplier.product?.discard()
productSupplier.supplier?.discard()
```
and none changes would be persisted to product and supplier :slightly_smiling_face: 

I wouldn't agree that this makes no difference between using the command class and domain class for binding, because what I meant by "updating some additional properties" I meant rather those, that are not meant to be touched during the saving details process, e.g. binding `productSupplierPreferences`, `productPackages`, which would be possible if we used domain class, and is not possible with the command I've created.

File: grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: Regarding your first comment, we should continue to use the JSON marshaller framework (including marshaller registration in BootStrap.groovy) until we have migrated to JSON views.

Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy:38]: Regarding the second one, I understand your point. We just have a convention of using the name action<DomainClass>, like saveProduct, saveProductSupplier. 

We haven't really discussed having method names that imply scope. And I would say that the scope should be explicitly defined through better security measures, not though naming. But if we need 

If we want to adopt a new naming convention we should create a proposal on the new naming convention and agree on it before applying it. 

Here's an example of what a naming convention might look like. I don't like all of these convention, but I like the general idea of defining each convention, providing some justification and demonstration through a code example.
https://sergiodelamo.com/blog/grails-extra-conventions.html

File: src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy
Comment (jmiranda) [reply src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy:38]: @awalkowiak Regarding your comment about shared constaints, which class should be the single source of truth.

Comment (awalkowiak) [reply src/main/groovy/org/pih/warehouse/product/ProductSupplierDetailsCommand.groovy:38]: I'm guessing the domain in this case, and then we could override it if needed.

**Comment by jmiranda** [comment]
@kchelstowski @awalkowiak i'll try to go through each discussion and determine whether it should be dealt with now or later.

File: grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy
Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:367]: @awalkowiak @kchelstowski I'll let you decide how you want to handle this 
* keep separate save and update methods
* use a single save method

File: grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy
Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy:385]: I'd like to keep the code below if possible i.e. revert the assignSourceCode. I like the idea of abstracting the generation and assignment, but I'm not yet comfortable with how it's done in #4486 so we can postpone that for now.

**Comment by jmiranda** [comment]
> An easy way to prevent from updating them would be to either create some generic method, or just write those two lines manually after binding to the product suppleier domain:
```
productSupplier.properties = command.properties
productSupplier.product?.discard()
productSupplier.supplier?.discard()
```
I understand your point, but this is not a sustainable approach. You need to do this everywhere you're using the "productSupplier.properties = ... " data binding mechanism. And if you add a new association to ProductSupplier, you need to remember to go back and discard the association. 

We either need to make this "discard" mechanism to be more automatic (i.e. annotation on the service method?) OR better yet we need the protection to happen during data binding phase in the controller. Either a feature within the @BindUsing annotation or explicit whitelisting properties (i.e. bindData(object, includes: [])) the client is allowed to update for a given service method. Or some other mechanism.
```
bindData(productSupplier, command.properties, [include: ['id', 'code', 'name', ...]])
```
We could potentially make this more palatable and DRY by create a command class interface to requires us to expose the properties to include and exclude via the command class.
```
bindData(productSupplier, command.properties, [include: command.includeProperties, exclude: command.excludeProperties])
```
> So I am not surprised that we can edit the Product and Organization, because those are existing domain object that are bound to the transaction/session and any changes on them would be tracked.

The point was that if you're binding objects to a domain or command, you can edit the properties of any object in their graph linked to that object. In order for command objects to be used to prevent this then your command object needs to bind only primitives and Strings (i.e. the PK of the object rather than the object itself).
```
String productId
String supplierId
``` 
Which, in my mind, defeats the purpose of using a command object for data binding. So again, there needs to be another way to protect against deep data binding. 
```
// This is a hypothetical for what i'm looking for
@BindUsing(onlyBindPrimaryKey=true)
Product product
```
However, I'm actually wondering whether the use of primitives + String in the command class would actually be all we need. 
```
productSupplier.properties = command.properties
```
It would be interesting to see if this approach yields what we're looking to do. It doesn't provide all of the validation we need at the command object level (i.e. we can have a productId in the command, but it might not actually map to a Product in the database), but that would be caught when we try to validate and save the domain. 

Anyway, that would be fine with me for now.

**Comment by kchelstowski** [comment]
@jmiranda I throw one more possible option to avoid updating nested properties, e.g. we can just define a proper validator in the command's constraints and see if a property (let's say - product) is dirty:
```groovy
 static constraints = {
      ...
      product(validator: { Product product ->
          if (product?.dirty) {
              return ['validator.product.cantBeDirty']
          }
          return true
      })
  }

**Comment by jmiranda** [comment]
Ooh dang. That's interesting. It feels a bit heavy-handed (a slap instead of an ignore) and it requires maintainance for whenever we add new associations, but I think it's an interesting avenue to explore, especially if we can make it a little more reusable. 

ProductSupplierCommand.groovy
```
static constraints = { 
    product: shared: 'myFavoriteLittleDirtyConstraint'
    supplier: shared: 'myFavoriteLittleDirtyConstraint'
}
```

runtime.groovy
```
grails.gorm.default.constraints = {
  myFavoriteLittleDirtyConstraint = { value, object ->
     if (value?.dirty) {
         // TODO exercise for reader: return a specific error for each class (if it exists), generic error message if not
         return ['validator.myFavoriteLittleDirtyConstraint.message']
     }
     return true
  }
}
```

> but had no success in that.

Can you clarify what you meant by that? The product wasn't bound at all or wasn't read-only and still bound properties we didn't want? Remember the "read" doesn't make the object read-only as-in unmodifiable, it just disables the dirty detection and presumably automatic flush at session end.

Source: https://docs.grails.org/latest/ref/Domain%20Classes/read.html

However, this is the path I'd like to explore a little more. I'm guessing we can do some interesting things with the BindUsing annotations.
```
@BindUsing({ obj, source -> Product.read(source["id"])})
Product product
```
If not, we can explore creating our own annotation or extend the existing one to do what we want i.e. discard object.

**Comment by kchelstowski** [comment]
> Can you clarify what you meant by that? The product wasn't bound at all or wasn't read-only and still bound properties we didn't want? 

Was not bound at all

**Comment by kchelstowski** [comment]
@awalkowiak @jmiranda I've addressed suggestions marked with " :+1:  "

**Review by awalkowiak — APPROVED** [review]
I think all of the change requests were resolved. LGTM

**Pull request merged**
